docs/en/msaicerr/environment_check.md
# Checking the Environment

## Description

Run a built-in operator sample to check the software and hardware environments.

## Command Format

```
python3 msaicerr.py -e -dev 0
```

## Parameters

-   **-e**  or  **--env**: The parameter is mandatory. It indicates checking the environment.
-   **-dev**  or  **--device\_id**: The parameter is optional. It specifies the ID of the device where the built-in operator sample runs. If this parameter is not set, the default device ID is  **0**. The msaicerr tool runs a built-in operator sample to check whether the software and hardware environments are normal.

## Command Example and Output Description

```
python3 msaicerr.py -e
```

Output example:

```
[INFO] Total device count: 1
[INFO] Valid device_id 0
[INFO] Get soc_version: xxxxxxx
[INFO] Start to test env with golden op.
[INFO] The built-in sample operator runs successfully, The environment is normal.
```

After the msaicerr.py tool is executed, the  **debug\_info.txt**  file is generated in the same directory as the msaicerr.py tool to record the log information generated during the tool execution.


docs/zh/env-vars/PROFILING_OPTIONS.md
# PROFILING\_OPTIONS

## 功能描述

训练或在线推理场景下，开发者可通过PROFILING\_OPTIONS环境变量配置Profiling配置选项。

### output

Profiling采集结果文件保存路径。支持配置绝对路径或相对路径（相对执行命令行时的当前路径）。路径中不能包含特殊字符："\\n"、"\\f"、"\\r"、"\\b"、"\\t"、"\\v"、"\\u007F"。

- 绝对路径配置以“/”开头，例如：/home/output。
- 相对路径配置直接以目录名开始，例如：output。
- 该参数优先级高于ASCEND\_WORK\_PATH。
- 该路径无需用户提前创建，采集过程中会自动创建。

### storage\_limit

指定落盘目录允许存放的最大文件容量。当Profiling数据文件在磁盘中即将占满本参数设置的最大存储空间或剩余磁盘总空间即将被占满时（总空间剩余<=20MB），则将磁盘内最早的文件进行老化删除处理。
<br>范围\[200, 4294967295\]，单位为MB，设置该参数时必须带单位，例如200MB。
<br>未配置本参数时，采集前，如果磁盘可用空间小于20MB时，则不落盘数据。

### training\_trace

采集迭代轨迹数据开关，即训练任务及AI软件栈的软件信息，实现对训练任务的性能分析，重点关注前后向计算和梯度聚合更新等相关数据。当采集正向和反向算子数据时该参数必须配置为on。

### task\_trace、task\_time

控制采集算子下发耗时和算子执行耗时的开关。涉及在task\_time、op\_summary、op\_statistic等文件中输出相关耗时数据。配置值：

- on：开启，默认值，和配置为l1的效果一样。
- off：关闭。
- l0：采集算子下发耗时、算子执行耗时数据。与l1相比，由于不采集算子基本信息数据，采集时性能开销较小，可更精准统计相关耗时数据。
- l1：采集算子下发耗时、算子执行耗时数据以及算子基本信息数据，提供更全面的性能分析数据。

当训练profiling mode开启即采集训练Profiling数据时，配置task\_trace为on的同时training\_trace也必须配置为on。

### ge\_api

采集动态shape算子在Host调度阶段的耗时数据。取值：

- off：关闭，默认为off。
- l0：采集动态shape算子在Host调度主要阶段的耗时数据，可更精准统计相关耗时数据。
- l1：采集动态shape算子在Host调度阶段更细粒度的耗时数据，提供更全面的性能分析数据。

### hccl

控制通信数据采集开关，可选on或off，默认为off。

> [!NOTE]说明
>此开关后续版本会废弃，请使用task\_time开关控制相关数据采集。

### aicpu

采集AICPU算子的详细信息，如：算子执行时间、数据拷贝时间等。取值on/off，默认为off。

### fp\_point

指定训练网络迭代轨迹正向算子的开始位置，用于记录前向计算开始时间戳。配置值为指定的正向第一个算子名字。用户可以在训练脚本中，通过tf.io.write\_graph将graph保存成.pbtxt文件，并获取文件中的name名称填入；也可直接配置为空，由系统自动识别正向算子的开始位置，例如"fp\_point":""。

### bp\_point

指定训练网络迭代轨迹反向算子的结束位置，记录后向计算结束时间戳，bp\_point和fp\_point可以计算出正反向时间。配置值为指定的反向最后一个算子名字。用户可以在训练脚本中，通过tf.io.write\_graph将graph保存成.pbtxt文件，并获取文件中的name名称填入；也可直接配置为空，由系统自动识别反向算子的结束位置，例如"bp\_point":""。

### aic\_metrics

AI Core性能指标采集项，取值如下：

- ArithmeticUtilization：各种计算类指标占比统计。
- PipeUtilization：计算单元和搬运单元耗时占比，该项为默认值。
- Memory：外部内存读写类指令占比。
- MemoryL0：内部L0内存读写类指令占比。
- MemoryUB：内部UB内存读写类指令占比。
- ResourceConflictRatio：流水线队列类指令占比。
- L2Cache：读写L2 Cache命中次数和缺失后重新分配次数。

  <!-- npu="310p" id6 -->
  Atlas推理系列产品不支持该参数。
  <!-- end id6 -->

  <!-- npu="910" id7 -->
  Atlas训练系列产品不支持该参数。
  <!-- end id7 -->
- MemoryAccess：算子在核上访存的带宽数据量。

  <!-- npu="310p" id8 -->
  Atlas推理系列产品不支持该参数。
  <!-- end id8 -->

  <!-- npu="910" id9 -->
  Atlas训练系列产品不支持该参数。
  <!-- end id9 -->

  <!-- npu="950" id10 -->
  Ascend 950PR&950DT系列产品：不支持该参数。
  <!-- end id10 -->

  > [!NOTE]说明
  >支持自定义需要采集的寄存器，例如："aic\_metrics":"**Custom:**_0x49,0x8,0x15,0x1b,0x64,0x10_"。
  >- Custom字段表示自定义类型，配置为具体的寄存器值，范围\[0x1, 0x6E\]。
  >- 配置的寄存器数最多不能超过8个，寄存器通过“,”区分开。
  >- 寄存器的值支持十六进制或十进制。

### l2

控制L2 Cache和TLB页表缓存命中率的开关，可选on或off，默认为off。若在aclgraph场景执行模型阶段开启Profiling，则该采集项无法生效。

<!-- npu="310p" id11 -->
- Atlas推理系列产品：支持采集L2 Cache的命中率。
<!-- end id11 -->

<!-- npu="910" id12 -->
- Atlas训练系列产品：支持采集L2 Cache的命中率。
<!-- end id12 -->

<!-- npu="910b" id13 -->
- Atlas A2系列产品品：支持采集L2 Cache和TLB页表缓存的命中率；分析AI Core命中L2次数推荐使用aic-metrics=L2Cache。
<!-- end id13 -->

<!-- npu="A3" id14 -->
- Atlas A3系列产品：支持采集L2 Cache和TLB页表缓存的命中率；分析AI Core命中L2次数推荐使用aic-metrics=L2Cache。
<!-- end id14 -->

<!-- npu="950" id15 -->
- Ascend 950PR&950DT系列产品：支持采集L2 Cache和TLB页表缓存的命中率；分析AI Core命中L2次数推荐使用aic-metrics=L2Cache。
<!-- end id15 -->

### msproftx

控制msproftx用户和上层框架程序输出性能数据的开关，可选on或off，默认值为off。
<br>需要先在应用程序脚本中添加如下mstx API或msproftx API，推荐使用mstx API。

### runtime\_api

控制runtime API性能数据采集开关，可选on或off，默认为off。可采集runtime API性能数据，包括Host与Device之间、Device间的同步异步内存复制时延等。

### sys\_hardware\_mem\_freq

片上内存、QoS传输带宽、LLC三级缓存带宽、加速器带宽、SoC传输带宽、组件内存占用等的采集开关。不同产品的采集内容略有差异，请以实际结果为准。范围\[1,100\]，单位Hz。

<!-- npu="950" id16 -->
Ascend 950PR&950DT系列产品，Qos和SoC支持的采集频率最大支持配置10000，其他采集项支持的最大采集频率仍为100，若配置超出范围，其他采集项则按照最大采集频率100进行采集。
<!-- end id16 -->

已知在安装有glibc<2.34的环境上采集memory数据，可能触发glibc的一个已知[Bug 19329](https://sourceware.org/bugzilla/show_bug.cgi?id=19329)，通过升级环境的glibc版本可解决此问题。

<!-- npu="A3,910b,310b" id17 -->
对于以下型号，采集任务结束后，不建议用户增大采集频率，否则可能导致SoC传输带宽数据丢失。
<!-- end id17 -->

<!-- npu="310b" id18 -->
Atlas 200I/500 A2推理产品
<!-- end id18 -->

<!-- npu="910b" id19 -->
Atlas A2系列产品品
<!-- end id19 -->

<!-- npu="A3" id20 -->
Atlas A3系列产品
<!-- end id20 -->

### llc\_profiling

LLC Profiling采集事件，可以设置为：

- read：读事件，三级缓存读速率，默认为read。
- write：写事件，三级缓存写速率。

### sys\_io\_sampling\_freq

NIC、ROCE、UB带宽数据采集频率。范围\[1,100\]，单位Hz。

<!-- npu="310p" id21 -->
Atlas推理系列产品：不支持该参数。
<!-- end id21 -->

<!-- npu="910b" id22 -->
Atlas A2系列产品品：支持采集NIC、ROCE。
<!-- end id22 -->

<!-- npu="A3" id23 -->
Atlas A3系列产品：支持采集NIC、ROCE。
<!-- end id23 -->

<!-- npu="950" id24 -->
Ascend 950PR&950DT系列产品：支持采集UB带宽数据。
<!-- end id24 -->

### sys\_interconnection\_freq

集合通信带宽数据（HCCS）、集合通信硬件加速单元（CCU）带宽数据、SIO数据、PCIe数据、UB带宽数据采集频率以及片间传输带宽信息采集频率。范围\[1,50\]，单位Hz。

<!-- npu="910" id25 -->
- Atlas训练系列产品：支持采集HCCS、PCIe数据。
<!-- end id25 -->

<!-- npu="910b" id26 -->
- Atlas A2系列产品品：支持采集HCCS、PCIe数据、片间传输带宽信息。
<!-- end id26 -->

<!-- npu="A3" id27 -->
- Atlas A3系列产品：支持采集HCCS、PCIe数据、片间传输带宽信息、SIO数据。
<!-- end id27 -->

<!-- npu="950" id28 -->
- Ascend 950PR&950DT系列产品：支持采集PCIe数据、片间传输带宽信息、CCU带宽数据、SIO数据、UB带宽数据。
<!-- end id28 -->

### dvpp\_freq

DVPP采集频率。范围\[1,100\]，单位Hz。

### instr\_profiling

AI Core和AI Vector的带宽和延时采集开关。仅单算子场景支持。取值on/off，默认为off。

<!-- npu="910" id29 -->
- Atlas训练系列产品：不支持该功能。
<!-- end id29 -->

<!-- npu="910b" id30 -->
- Atlas A2系列产品品：不支持该开关，通过instr\_profiling\_freq控制该功能。
<!-- end id30 -->

<!-- npu="A3" id31 -->
- Atlas A3系列产品：不支持该开关，通过instr\_profiling\_freq控制该功能。
<!-- end id31 -->

<!-- npu="950" id32 -->
- Ascend 950PR&950DT系列产品：支持，但可能会因最后一段指令的统计时间超长导致统计不准确，建议使用msprof op方式采集。
<!-- end id32 -->

### instr\_profiling\_freq

AI Core和AI Vector的带宽和延时采集开关，配置了采集频率即开启相关采集能力。仅单算子场景支持。范围\[300,30000\]，单位Hz。

<!-- npu="910" id33 -->
- Atlas训练系列产品：不支持该功能。
<!-- end id33 -->

<!-- npu="910b" id34 -->
- Atlas A2系列产品品：支持。
<!-- end id34 -->

<!-- npu="A3" id35 -->
- Atlas A3系列产品：支持。
<!-- end id35 -->

<!-- npu="950" id36 -->
- Ascend 950PR&950DT系列产品：不支持该开关，通过instr\_profiling控制该功能。
<!-- end id36 -->

### host\_sys

Host侧性能数据采集开关。取值如下，可选其中的一项或多项，选多项时用英文逗号隔开，例如"host\_sys": "cpu,mem"。

- cpu：进程级别的CPU利用率。
- mem：进程级别的内存利用率。
- disk：进程级别的磁盘I/O利用率。采集Host侧disk性能数据需要安装第三方开源工具iotop，采集osrt性能数据需要安装第三方开源工具perf和ltrace。
- network：系统级别的网络I/O利用率。
- osrt：进程级别的syscall和pthreadcall。

### host\_sys\_usage

采集Host侧系统及所有进程的CPU和内存数据。取值包括cpu和mem，可选其中的一项或多项，选多项时用英文逗号隔开。

### host\_sys\_usage\_freq

配置Host侧系统和所有进程CPU、内存数据的采集频率。范围\[1,50\]，默认值50，单位Hz。

> [!NOTE]说明
>
>- 除动态shape场景外的其他场景，fp\_point、bp\_point为自动配置项，无需用户手动配置。动态shape场景不支持自动配置fp/bp，需要用户手动设置。
>- 在线推理支持task\_trace和aicpu，不支持training\_trace。

## 配置示例

```sh
export PROFILING_OPTIONS='{"output":"/tmp/profiling","training_trace":"on","task_trace":"on","fp_point":"","bp_point":"","aic_metrics":"PipeUtilization"}'
```

## 使用约束

- 此环境变量仅适用于TensorFlow训练或在线推理场景。
- 使用此环境变量前，请通过环境变量PROFILING\_MODE开启Profiling采集功能。

## 支持的型号

<!-- npu="910" id1 -->
Atlas训练系列产品
<!-- end id1 -->

<!-- npu="310p" id2 -->
Atlas推理系列产品
<!-- end id2 -->

<!-- npu="910b" id3 -->
Atlas A2系列产品品
<!-- end id3 -->

<!-- npu="A3" id4 -->
Atlas A3系列产品
<!-- end id4 -->

<!-- npu="950" id5 -->
Ascend 950PR&950DT系列产品
<!-- end id5 -->


src/msprof/collector/dvvp/common/config/config.h
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef ANALYSIS_DVVP_COMMON_CONFIG_CONFIG_H
#define ANALYSIS_DVVP_COMMON_CONFIG_CONFIG_H

#include <cstdint>
#include <string>
#include <map>

namespace analysis {
namespace dvvp {
namespace common {
namespace config {
// /////////////////////common//////////////////////////////////
const std::string DEVICE_APP_DIR = "/usr/local/profiler/";
const std::string PROF_SCRIPT_FILE_PATH = "/usr/bin/msprof_data_collection.sh";
const std::string PROF_SCRIPT_PROF = "/var/prof_collect.sh";
const std::string SAMPLE_JSON = "sample.json";

const std::string ASCEND_HOSTPID = "ASCEND_HOSTPID";
const std::string ASCEND_WORK_PATH_ENV = "ASCEND_WORK_PATH";
const std::string PROFILING_RESULT_PATH = "profiling_data";

const std::string CONTAINER_NO_SUPPORT_MESSAGE = "MESSAGE_CONTAINER_NO_SUPPORT";

const char* const DEVICE_TAG_KEY = "Device";
const char* const CLOCK_MONOTONIC_RAW_KEY = "clock_monotonic_raw";
const char* const CLOCK_CNTVCT_KEY = "cntvct";
const char* const CLOCK_CNTVCT_KEY_DIFF = "cntvct_diff";

constexpr int32_t US_CONVERT_MS = 1000;
constexpr int32_t HZ_CONVERT_MS = 1000;
constexpr int32_t HZ_CONVERT_US = 1000000;
constexpr int32_t HZ_HUNDRED = 100;
constexpr int32_t HZ_TEN_THOUSAND = 10000;

constexpr int32_t MAX_PATH_LENGTH = 1024;

constexpr int32_t PROFILING_PACKET_MAX_LEN = (3 * 1024 * 1024); // 3 * 1024 *1024 means 3mb

constexpr int32_t MSVP_BATCH_MAX_LEN = 2621440; // 2621440 : 2.5 * 1024 *1024 means 2.5MB

constexpr int32_t RECEIVE_CHUNK_SIZE = 320; // chunk size:1024

constexpr int32_t HASH_DATA_MAX_LEN = 1024; // hash data max len:1024

constexpr int32_t MSVP_COLLECT_CLIENT_INPUT_MAX_LEN = 128; // 128 is max string size

constexpr int32_t MSVP_PROFILER_THREADNAME_MAXNUM = 16;

constexpr int32_t MSVP_DECODE_MESSAGE_MAX_LEN = (128 * 1024 * 1024); // 128 * 1024 *1024 means 128mb

constexpr int32_t MSVP_MESSAGE_TYPE_NAME_MAX_LEN = 1024; // 1024 means 1KB

constexpr int32_t MSVP_CFG_MAX_SIZE = 64 * 1024 * 1024; // 64 * 1024 *1024 means 64mb

constexpr long long MSVP_LARGE_FILE_MAX_LEN = 512 * 1024 * 1024; // 512 * 1024 * 1024 means 512mb
constexpr long long MSVP_SMALL_FILE_MAX_LEN = 2 * 1024 * 1024;   // 2 * 1024 * 1024 means 2mb
constexpr uint32_t HOST_DATA_BUF_SIZE_M = 2097153;               // 2097152 + 1 means 2m + '\0'

constexpr int32_t MSVP_CLN_SENDER_QUEUE_CAPCITY = 1024;
constexpr int32_t MSVP_CLN_SENDER_POOL_THREAD_NUM = 2;

constexpr uint64_t STORAGE_LIMIT_DOWN_THD = 200; // 200MB

constexpr int32_t MAX_PMU_EVENT = 2147483647;     // max ai core event is now 0x7FFFFFFF
constexpr int32_t LITE_MAX_PMU_EVENT = 1813;      // max ai core event is 0x715 for chip:12
constexpr int32_t ACC_MAX_PMU_EVENT = 2147483647; // max ai core event is 0x7FFFFFFF for chip:15

constexpr int32_t MAX_PARAMS_LEN = 70;
constexpr int32_t MAX_APP_LEN = 2048;
constexpr int32_t HEX_MODE = 16;
constexpr int32_t DEC_MODE = 10;
constexpr int32_t MAX_PID_LENGTH = 8;
const std::string CUSTOM_METRICS = "Custom:";
const std::string NULL_CHUNK = "null";

constexpr uint32_t MS_TO_NS = 1000000; // 1ms = 1000000ns

#if (defined(_WIN32) || defined(_WIN64) || defined(_MSC_VER))
constexpr int32_t MSVP_PROCESS = NULL;
#else
constexpr int32_t MSVP_PROCESS = -1;
#endif

const char* const MSVP_UNDERLINE = "_";

constexpr int32_t HOST_PID_DEFAULT = -1;

const char* const MSVP_CHANNEL_POOL_NAME_PREFIX = "MSVP_ChanPool_";
const char* const MSVP_SENDER_POOL_NAME_PREFIX = "MSVP_SndPool_";

const char* const MSVP_CHANNEL_THREAD_NAME = "MSVP_ChanPoll";
const char* const MSVP_CHANNEL_BUFFER = "MSVP_ChanBuffer";
const char* const MSVP_CTRL_RECEIVER_THREAD_NAME = "MSVP_CtrlRecv";
const char* const MSVP_UPLOADER_THREAD_NAME = "MSVP_Upld";
const char* const MSVP_DEVICE_TRANSPORT_THREAD_NAME = "MSVP_DevTrans";
const char* const MSVP_PROF_TASK_THREAD_NAME = "MSVP_ProfTask";
const char* const MSVP_DEVICE_THREAD_NAME_PREFIX = "MSVP_Dev_";
const char* const MSVP_COLLECT_PERF_SCRIPT_THREAD_NAME = "MSVP_PerfScript";
const char* const MSVP_COLLECT_PROF_TIMER_THREAD_NAME = "MSVP_ProfTimer";
const char* const MSVP_UPLOADER_DUMPER_THREAD_NAME = "MSVP_UploaderDumper";
const char* const MSVP_HDC_DUMPER_THREAD_NAME = "MSVP_HdcDumper";
const char* const MSVP_RPC_DUMPER_THREAD_NAME = "MSVP_RpcDumper";
const char* const MSVP_HELPER_DUMPER_THREAD_NAME = "MSVP_HelperDumper";
const char* const MSVP_DYN_PROF_SERVER_THREAD_NAME = "MSVP_DynProfServer";
const char* const MSVP_DYN_PROF_CLIENT_THREAD_NAME = "MSVP_DynProfClient";
const char* const MSVP_TYPE_INFO_UPLOAD_THREAD_NAME = "MSVP_TypeInfoUpload";
const char* const MSVP_MSTX_DATA_HANDLE_THREAD_NAME = "MSVP_MstxDataHandle";
const char* const MSVP_DIAGNOSTIC_THREAD_NAME = "MSVP_Diagnostic";

// cloud prof config
constexpr int32_t MAX_DEVICE_NUMS = 8;

constexpr size_t THREAD_QUEUE_SIZE_DEFAULT = 64;

// prof peripheral job config
constexpr uint32_t DEFAULT_INTERVAL = 100;
constexpr uint32_t DEAFULT_MASTER_ID = 0xFFFFFFFF;
constexpr uint32_t PERIPHERAL_EVENT_READ = 0;
constexpr uint32_t PERIPHERAL_EVENT_WRITE = 1;
constexpr uint32_t PERIPHERAL_EVENT_APP_MEM = 0;
constexpr uint32_t PERIPHERAL_EVENT_DEV_MEM = 1;
constexpr int32_t PERIPHERAL_INTERVAL_MS_SMIN = 10;
constexpr int32_t PERIPHERAL_INTERVAL_MS_MIN = 1;
constexpr int32_t PERIPHERAL_INTERVAL_MS_MAX = 1000;

// prof job config
const char* const PROF_SYS_CPU_USAGE_FILE = "SystemCpuUsage.data";
const char* const PROF_SYS_MEM_FILE = "Memory.data";
const char* const PROF_HOST_PROC_CPU_USAGE_FILE = "host_cpu.data";
const char* const PROF_HOST_CPU_FREQ_FILE = "host_cpu_freq.data";
const char* const PROF_HOST_PROC_MEM_USAGE_FILE = "host_mem.data";
const char* const PROF_HOST_SYS_NETWORK_USAGE_FILE = "host_network.data";
const char* const PROF_NETDEV_STATS_FILE = "netdev_stats.data";
const char* const MSVP_PROF_DATA_DIR = "/data";
const char* const MSVP_PROF_PERF_DATA_FILE = "ai_ctrl_cpu.data.";
const char* const MSVP_PROF_PERF_RET_FILE_SUFFIX = ".txt";
const std::string PROF_AICORE_SAMPLE = "aicore sample based";
const std::string PROF_AIV_SAMPLE = "ai vector core sample based";
const std::string PROF_AICORE_TASK = "aicore task based";
const std::string PROF_AIV_TASK = "ai vector core task based";
const std::string PROF_HOST_JOBID = "64";

// prof task config
constexpr unsigned long PROCESS_WAIT_TIME = 500000; // should not modify
const std::string PROF_TASK_STREAMTASK_QUEUE_NAME = "ProfTaskStreamBuffer";

// uploader config
const std::string UPLOADER_QUEUE_NAME = "UploaderQueue";
constexpr size_t UPLOADER_QUEUE_CAPACITY = 512;

// prof channel config
const std::string SPEED_PERFCOUNT_MODULE_NAME = std::string("ChannelReaderSpeed");
const std::string SPEEDALL_PERFCOUNT_MODULE_NAME = std::string("ChannelReaderSpeedAll");
constexpr size_t CHANNELPOLL_THREAD_QUEUE_SIZE = 8192;

// prof mgr config
const std::string PROF_FEATURE_TASK = "task_trace";
constexpr int32_t PROF_MGR_TRACE_ID_DEFAULT_LEN = 27;

// receive data config
constexpr size_t MAX_LOOP_TIMES = 1400;                     // the max send package nums of once Dump()
constexpr int32_t SLEEP_INTEVAL_US = 1000;                  // the interval of Run()
constexpr size_t RING_BUFF_CAPACITY = 32768;                // 32768:32K. Note:capacity value must be 2^n
constexpr size_t COMPACT_RING_BUFF_CAPACITY = 65536;        // 65536:64K. Note:capacity value must be 2^n
constexpr size_t ADDITIONAL_RING_BUFF_CAPACITY = 131072;    // 131072:128K. Note:capacity value must be 2^n
constexpr size_t ADPROF_BUFF_CAPACITY = 16384;              // 16384:16K. Note:capacity value must be 2^n
constexpr size_t GE_RING_BUFF_CAPACITY = 262144;            // 262144:256K. Note:capacity value must be 2^n
constexpr size_t PROFTX_RING_BUFF_CAPACITY = 262144;        // 262144:256K. Note:capacity value must be 2^n
constexpr size_t VARIABLE_ADDITIONAL_BUFF_CAPACITY = 65536; // 65536:64K. Note:capacity value must be 2^n

// sender config
constexpr int32_t SEND_BUFFER_LEN = 64 * 1024; // 64 * 1024 means 64k
constexpr size_t SENDERPOOL_THREAD_QUEUE_SIZE = 8192;

// transport config config
const std::string HDC_PERFCOUNT_MODULE_NAME = std::string("HdcTransport");
const std::string FILE_PERFCOUNT_MODULE_NAME = std::string("FileTransport");
constexpr uint64_t TRANSPORT_PRI_FREQ = 128;

const char* const MSVP_PROF_ON = "on";
const char* const MSVP_PROF_OFF = "off";
const char* const MSVP_PROF_L0 = "l0";
const char* const MSVP_PROF_L1 = "l1";
const char* const MSVP_PROF_L2 = "l2";
const char* const MSVP_PROF_L3 = "l3";
const char* const MSVP_LEVEL_L0 = "level0";
const char* const MSVP_LEVEL_L1 = "level1";
const char* const MSVP_LEVEL_L2 = "level2";
const char* const MSVP_LEVEL_L3 = "level3";
const char* const MSVP_PROF_EMPTY_STRING = "";
const char* const MSVP_PROF_ACLAPI_MODE = "aclapi";
const char* const MSVP_PROF_SUBSCRIBE_MODE = "subscribe";
const char* const MSVP_PROF_SYSTEM_MODE = "system";
const char* const HELPER_HOST_CPU_MODE = "64";

// dynamic profiling
const std::string PROFILING_MODE_ENV = "PROFILING_MODE";
const std::string DYNAMIC_PROFILING_KEY_PID_ENV = "DYNAMIC_PROFILING_KEY_PID";
const std::string DYNAMIC_PROFILING_VALUE = "dynamic";
const std::string DELAY_DURARION_PROFILING_VALUE = "delay_or_duration";

// ai core metrics type
constexpr char ARITHMETIC_UTILIZATION[] = "ArithmeticUtilization";
constexpr char PIPE_UTILIZATION[] = "PipeUtilization";
constexpr char PIPE_UTILIZATION_EXCT[] = "PipeUtilizationExct";
constexpr char PIPE_EXECUTION_UTILIZATION[] = "PipelineExecuteUtilization";
constexpr char MEMORY_BANDWIDTH[] = "Memory";
constexpr char L0B_AND_WIDTH[] = "MemoryL0";
constexpr char RESOURCE_CONFLICT_RATIO[] = "ResourceConflictRatio";
constexpr char PROFILER_SAMPLE_CONFIG_ENV[] = "PROFILER_SAMPLECONFIG";
constexpr char MEMORY_UB[] = "MemoryUB";
constexpr char L2_CACHE[] = "L2Cache";
constexpr char MEMORY_ACCESS[] = "MemoryAccess";
constexpr char PIPESTALLCYCLE[] = "PipeStallCycle";
constexpr char SCALAR_RATIO[] = "ScalarRatio";
constexpr char RANK_ID_ENV[] = "RANK_ID";

// llc  profiling events type
constexpr char LLC_PROFILING_CAPACITY[] = "capacity";
constexpr char LLC_PROFILING_BANDWIDTH[] = "bandwidth";
constexpr char LLC_PROFILING_READ[] = "read";
constexpr char LLC_PROFILING_WRITE[] = "write";

// host sys events type
constexpr char HOST_SYS_CPU[] = "cpu";
constexpr char HOST_SYS_MEM[] = "mem";
constexpr char HOST_SYS_DISK[] = "disk";
constexpr char HOST_SYS_NETWORK[] = "network";
constexpr char HOST_SYS_OSRT[] = "osrt"; // os_runtime: system call && pthread
constexpr char HOST_SYS_NUMA[] = "numa";

constexpr char OUTPUT_RECORD[] = "profiling_output_record";

// keypoint op's name/type
const char* const KEYPOINT_OP_NAME = "keypoint_op";
const char* const KEYPOINT_OP_TYPE = "na";

// hash data file tag
const std::string HASH_TAG = "hash_dic";
const char* const HASH_DIC_DELIMITER = ":";

// need paired ageing file
const char* const HWTS_DATA = "hwts.data";
const char* const AICORE_DATA = "aicore.data";
const char* const STORAGE_LIMIT_UNIT = "MB";

// analysis script param
constexpr char DEFAULT_INTERATION_ID[] = "-1";
constexpr char DEFAULT_MODEL_ID[] = "-1";
const char* const PROFILING_SUMMARY_FORMAT = "csv";
const char* const PROFILING_EXPORT_TYPE_TEXT = "text";

// used for init param
constexpr int32_t DEFAULT_PROFILING_INTERVAL_5MS = 5;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_10MS = 10;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_20MS = 20;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_50MS = 50;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_100MS = 100;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_1000MS = 1000; // 1000 cycle
constexpr int32_t DEFAULT_PROFILING_INTERVAL_100US = 100;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_10000US = 10000;
constexpr int32_t DEFAULT_PROFILING_INTERVAL_20000US = 20000;

enum FileChunkDataModule {
    PROFILING_DEFAULT_DATA_MODULE = 0,
    PROFILING_IS_FROM_MSPROF,
    PROFILING_IS_CTRL_DATA,
    PROFILING_IS_FROM_DEVICE,
    PROFILING_IS_FROM_MSPROF_DEVICE,
    PROFILING_IS_FROM_MSPROF_HOST,
    PROFILING_IS_FROM_INNER
};

enum MsprofReporterId { API_EVENT = 0, COMPACT = 1, ADDITIONAL = 2, ADPROF = 3, VARIABLE_ADDINFO = 4 };

// instr profiling perf
constexpr int32_t DAVID_DIE0_AICORE_NUM = 18;    // die 18
constexpr int32_t BIU_PERF_LOWER_GROUP_NUM = 3;  // biu profiling group id 0 ~ 2
constexpr int32_t BIU_PERF_HIGHER_GROUP_NUM = 6; // biu profiling group id 0 ~ 5
constexpr int32_t DEFAULT_BIU_PERF_CYCLE = 16;
constexpr int32_t INSTR_PROFILING_GROUP_MAX_NUM = 25;      // instr profiling group id 0 ~ 24
constexpr int32_t INSTR_PROFILING_GROUP_CHANNEL_NUM = 3;   // instr profiling group contains 3 channel
constexpr int32_t INSTR_PROFILING_SAMPLE_FREQ_MIN = 300;   // instr profiling sampling frequency min value
constexpr int32_t INSTR_PROFILING_SAMPLE_FREQ_MAX = 30000; // instr profiling sampling frequency max value

// number switch limit
constexpr uint32_t PROF_MAX_DYNAMIC_PID = 2147483647;
constexpr uint32_t PROF_MAX_DYNAMIC_TIME = UINT32_MAX;

constexpr uint32_t TLV_VALUE_MAX_LEN = 1200;
constexpr uint32_t TLV_VALUE_CHUNK_MAX_LEN = 1032;
constexpr uint32_t TLV_VALUE_FILENAME_MAX_LEN = 64;
constexpr uint32_t TLV_VALUE_EXTRAINFO_MAX_LEN = 64;
constexpr uint32_t TLV_VALUE_ID_MAX_LEN = 16;
constexpr uint32_t TLV_HEAD = 0x5A5A5A5AU;

struct ProfTlv {
    uint32_t head;
    uint32_t version;
    uint32_t type;
    uint32_t len;
    uint8_t value[TLV_VALUE_MAX_LEN];
};

struct ProfTlvValue {
    bool isLastChunk;
    int32_t chunkModule;
    size_t chunkSize;
    int64_t offset;
    char chunk[TLV_VALUE_CHUNK_MAX_LEN];
    char fileName[TLV_VALUE_FILENAME_MAX_LEN];
    char extraInfo[TLV_VALUE_EXTRAINFO_MAX_LEN];
    char id[TLV_VALUE_ID_MAX_LEN];
};

} // namespace config
} // namespace common
} // namespace dvvp
} // namespace analysis

#endif
